File: solution/c#/Day01/Day01/Food.cs
namespace Day01
{
    using DateProvider = Func<DateOnly>;

    public sealed record Food(
        DateOnly ExpirationDate,
        bool ApprovedForConsumption,
        Guid? InspectorId)
    {
        public bool IsEdible(DateProvider now)
            => IsFresh(now) &&
               CanBeConsumed() &&
               HasBeenInspected();

        private bool IsFresh(DateProvider now) => ExpirationDate.IsGreaterThan(now());
        private bool CanBeConsumed() => ApprovedForConsumption;
        private bool HasBeenInspected() => InspectorId != null;
    }
}

File: solution/ts/day01/src/food.ts
type Maybe<T> = NonNullable<T> | null;
type DateProvider = () => Date;

export class Food {
    constructor(
        public readonly expirationDate: Date,
        public readonly approvedForConsumption: boolean,
        public readonly inspectorId: Maybe<string>
    ) {
    }

    isEdible(now: DateProvider): boolean {
        return this.isFresh(now) &&
            this.canBeConsumed() &&
            this.hasBeenInspected();
    }

    private isFresh = (now: () => Date) => this.expirationDate > now();
    private canBeConsumed = () => this.approvedForConsumption;
    private hasBeenInspected = () => this.inspectorId !== null;
}